src/main/java/io/jenkins/plugins/appcenter/api/AppCenterServiceFactory.java

```diff
@@ -87,4 +109,26 @@ public UploadService createUploadService() {
     public Secret getApiToken() {
         return apiToken;
     }
+
+    private void setProxy(
+        final ProxyConfiguration cfg,
+        final OkHttpClient.Builder builder) {
+
+        builder
+            .proxy(new Proxy(HTTP, new InetSocketAddress(cfg.name, cfg.port)));
```

Comment:
There's a static method, `ProxyConfiguration::createProxy` that might help with this. You pass it in `(String host, String name, int port, String noProxyHost)` and if the host you are attempting to connect to is on the no proxy list it returns `Proxy.NO_PROXY` otherwise `new Proxy(Proxy.Type.HTTP, new InetSocketAddress(name,port))` as you've done here.

That's find for `api.appcenter.ms` but the upload url is returned from a previous api request. At the moment it will be of the form `rink.hockeyapp.net` but in the future I don't know what it might be (likely something in `api.appcenter.ms`) but who knows? [See this page](https://docs.microsoft.com/en-us/appcenter/distribution/uploading#uploading-using-the-apis).

So what we might have to do is delay calling `createUploadService` until we have the upload url, get the hostname, and attempt to set the proxy as above if it isn't on the no proxy list.

Reply:
I was actually surprised to see the upload URL `rink.hockeyapp.net` and not an appcenter domain. I think we will need to check for the hockeyapp domain being in the `NO_PROXY` config for now, then change it in a later version once the domain changes.

Reply:
Tried to follow your suggested approach, though I am unsure if I didn't overcomplicate things unnecessarily with my implementation. Maybe you have some suggestions to simplify it.